### PR TITLE
HERA-164 [Development][LMS part] - XBlock "Question" - Base frontend functionality.

### PR DESCRIPTION
**Pull request by UvgenGen**
[HERA-164](https://youtrack.raccoongang.com/issue/HERA-164) [Development][LMS part] - XBlock "Question" - Base frontend functionality.

### CODE REVIEW

File: hera_xblocks/question/question/question.py [hunk 1]
```diff
@@ -89,15 +116,44 @@ def student_view(self, context=None):
...
+        self.user_confidence = confidence
+
+    @XBlock.json_handler
+    def submit(self, answers, suffix=''):
+        user_answers = []
+        correct = False
+    
+        if self.question.get('type') == "number":
+            for answer in answers:
+                user_answer = int(answer.get('value'))
```
Comment (UvgenGen) [inline hera_xblocks/question/question/question.py]: add validation

File: hera_xblocks/question/question/question.py [hunk 2]
```diff
@@ -89,15 +116,44 @@ def student_view(self, context=None):
...
+            for answer in answers:
+                user_answers.append(answer.get('value'))
+                if answer.get('value') == self.answer:
+                    correct = True
+
+        elif self.question.get('type') in ["select", "radio", "checkbox"]:
+            correct_answers = [ option["title"] for option in self.options if option["correct"] is True ]
+            for answer in answers:
+                user_answers.append(answer.get('value'))
+            if user_answers == correct_answers:
```
Comment (UvgenGen) [inline hera_xblocks/question/question/question.py]: Use set for comparing

File: hera_xblocks/question/question/static/html/question.html [hunk 3]
```diff
@@ -1,20 +1,53 @@
...
+                {% elif question.type == "select" %}
+                    <select name="answer">
+                        <option value=""> ------- </option>
+                        {% for option in options %}
+                            <option value="{{option.title}}"{% if option.title in user_answer %} selected{% endif %}>{{option.title}}</option>
+                        {% endfor %}
+                    </select>
+                {% else %}
+                    {% for option in options %}
+                        <p><input type="{{question.type}}" name="answer" value="{{option.title}}"{% if option.title in user_answer %} checked{% endif %}> {{option.title}} <p>
```
Comment (UvgenGen) [inline hera_xblocks/question/question/static/html/question.html]: Line length

File: hera_xblocks/question/question/static/html/question.html [hunk 4]
```diff
@@ -1,20 +1,53 @@
...
+        <button type="button" class="author-block__btn hidden" id="skip_{{block_id}}">Skip</button>
+        <button type="button" class="author-block__btn" id="submit_{{block_id}}" disabled>Submit</button>
+    </div>
+    <div id="overlay">
+        <div id="popup">
+        <div class="close" id="iframe-close">&times;</div>
+        <iframe src="{{iframe_url}}" class="simulation-iframe"></iframe>
+        </div>
     </div>
 </div>
```
Comment (UvgenGen) [inline hera_xblocks/question/question/static/html/question.html]: empty line

File: hera_xblocks/question/question/static/html/question.html [hunk 5]
```diff
@@ -1,20 +1,55 @@
...
+                {% else %}
+                    {% for option in options %}
+                        <p>
+                        <input type="{{question.type}}" name="answer" value="{{option.title}}"
+                        {% if option.title in user_answer %} checked{% endif %}> {{option.title}} <p>
+                    {% endfor %}
+                {% endif %}
+            </form>
+            <div class="scaffold-info hidden">
+                <img class="scaffold-img" alt="scaffold img">
```
Comment (sendr) [inline hera_xblocks/question/question/static/html/question.html]: add src please

File: hera_xblocks/question/question/static/html/question.html [hunk 6]
```diff
@@ -1,20 +1,55 @@
...
+        <input type="number" class="confidence-input hidden" min="0" max="100">
+        <span class="confidence-text error">Please tell us how confident you are about your answer.</span>
+        <button type="button" class="author-block__btn hidden" id="skip_{{block_id}}">Skip</button>
+        <button type="button" class="author-block__btn" id="submit_{{block_id}}" disabled>Submit</button>
     </div>
-</div>
\ No newline at end of file
+    <div id="overlay">
+        <div id="popup">
+        <div class="close" id="iframe-close">&times;</div>
```
Comment (sendr) [inline hera_xblocks/question/question/static/html/question.html]: indentations please

File: hera_xblocks/question/question/question.py [hunk 7]
```diff
@@ -89,15 +119,45 @@ def student_view(self, context=None):
...
     @XBlock.json_handler
-    def submit(self, answer, suffix=''):
-        correct_answers = [ option["title"] for option in self.options if option["correct"]==True ]
-        print(answer, correct_answers)
-        return True if answer in correct_answers else False
+    def submit(self, data, suffix=''):
+        user_answers = []
+        correct = False
+        answers = data.get("answers")
+        self.user_confidence = int(data.get("confidence"))
```
Comment (sendr) [inline hera_xblocks/question/question/question.py]: where does a validation take place? what if confidence isn't integer?

File: hera_xblocks/question/question/static/html/question.html [hunk 8]
```diff
@@ -1,20 +1,60 @@
...
-        {{description|safe}}
-        <form id="question">
-        {% for option in options %}
-            <p><input type="radio" name="answer" value="{{option.title}}"> {{option.title}} </option><p>
-        {% endfor %}
-        </form>
+            <div class="question__content">{{description|safe}}</div>
+            <form class="question-form">
+                {% if question.type == "text" or question.type == "number" %}
+                    <input type="{{question.type}}" name="answer" value="{{user_answer|first}}">
```
Comment (sendr) [inline hera_xblocks/question/question/static/html/question.html]:
```suggestion
                    <input type="{{question.type}}" name="answer" value="{{user_answer|first}}"/>
```

File: hera_xblocks/question/question/static/html/question.html [hunk 9]
```diff
@@ -1,20 +1,60 @@
...
+                    </select>
+                {% else %}
+                    {% for option in options %}
+                        <p>
+                            <input
+                                type="{{question.type}}"
+                                name="answer"
+                                value="{{option.title}}"
+                                {% if option.title in user_answer %} checked{% endif %}
+                            > {{option.title}}
```
Comment (sendr) [inline hera_xblocks/question/question/static/html/question.html]:
```suggestion
                            /> {{option.title}}
```

File: hera_xblocks/question/question/static/html/question.html [hunk 10]
```diff
@@ -1,20 +1,60 @@
...
+                        <p>
```
Comment (sendr) [inline hera_xblocks/question/question/static/html/question.html]: the second opening tag `<p>`, where are the closing tags?

File: hera_xblocks/question/question/static/html/question.html [hunk 11]
```diff
@@ -1,20 +1,60 @@
...
+                <button type="button" class="scaffold" data-img_url="{{teach_me.imgUrl}}" data-content="{{teach_me.content}}">{{teach_me.content}}</button>
+            </div>
         </div>
     </div>
-    <div class="author-block__buttons"><button type="button" class="author-block__btn" id="submit">Submit</button></div>
-    <div id="iframe">
-        <iframe src="{{iframe_url}}" style="display: block; width: 600px; height: 600px;"></iframe>
+    <div class="author-holder author-block__buttons">
+        <span class="confidence-text info hidden">{{confidence_text}}:</span>
+        <input type="number" class="confidence-input hidden" min="0" max="100">
```
Comment (sendr) [inline hera_xblocks/question/question/static/html/question.html]:
```suggestion
        <input type="number" class="confidence-input hidden" min="0" max="100"/>
```

File: hera_xblocks/question/question/question.py [hunk 12]
```diff
@@ -87,17 +114,51 @@ def student_view(self, context=None):
...
+        elif self.question.get('type') == "text":
+                user_answers.append(answers.get('value'))
+                if answer.get('value') == self.answer:
+                    correct = True
+
+        elif self.question.get('type') in ["select", "radio", "checkbox"]:
+            correct_answers = [ option["title"] for option in self.options if option["correct"] is True ]
+            for answer in answers:
+                user_answers.append(answer.get('value'))
+            if set(user_answers) == set(correct_answers):
```
Comment (sendr) [inline hera_xblocks/question/question/question.py]:
```suggestion
            if set(answers) == set(correct_answers):
```
and we don't need the loop above?

File: hera_xblocks/question/question/question.py [hunk 13]
```diff
@@ -87,17 +114,51 @@ def student_view(self, context=None):
...
+        user_answers = []
+        correct = False
+        answers = data.get("answers")
+        try:
+            self.user_confidence = int(data.get("confidence"))
+        except ValueError:
+            self.user_confidence = None
+
+        if self.question.get('type') == "number":
+                try:
```
Comment (sendr) [inline hera_xblocks/question/question/question.py]: wrong indentation

File: hera_xblocks/question/question/question.py [hunk 14]
```diff
@@ -87,17 +114,51 @@ def student_view(self, context=None):
...
+    def submit(self, data, suffix=''):
+        user_answers = []
+        correct = False
+        answers = data.get("answers")
+        try:
+            self.user_confidence = int(data.get("confidence"))
+        except ValueError:
+            self.user_confidence = None
+
+        if self.question.get('type') == "number":
```
Comment (sendr) [inline hera_xblocks/question/question/question.py]: `self.question.get('questionType')`

File: hera_xblocks/question/question/question.py [hunk 15]
```diff
@@ -87,17 +114,51 @@ def student_view(self, context=None):
...
+                try:
+                    user_answer = float(answers.get('value'))
+                    user_answers.append(user_answer)
+                    correct_answer = self.answer
+                    if correct_answer - self.preciseness <= user_answer <= correct_answer + self.preciseness:
+                        correct = True
+                except ValueError:
+                    user_answer = None
+
+        elif self.question.get('type') == "text":
```
Comment (sendr) [inline hera_xblocks/question/question/question.py]: `self.question.get('questionType')`

**Pull request merged**